# Algoritmos de optimización - TRABAJO PRÁCTICO<br>
Nombre y Apellidos: Daniel Oriol Asensio  <br>
Link: https://colab.research.google.com/drive/1BXfv5qAtPalKB3Sy6AKnQG0Cd5z3V1yg#scrollTo=hVbXYX-RfPWh<br>
Url: https://github.com/doriola/03MIAR---Algoritmos-de-Optimizacion_2023.git<br>
Problema:
>3. Combinar cifras y operaciones

Descripción del problema:<br>

• El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.<br>
• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)<br>
• Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad dada.<br> 
Un ejemplo sería para obtener el 4:<br>
4+2-6/3*1 = 4

                                

In [23]:
import numpy as np
from itertools import *
import math
import random

In [2]:
# Genero una lista que contiene en cada elemento una lista con 5 números 
# correspondientes a las permutaciones posibles a tener en cuenta en la 
# solución.
num_comb=5
permutaciones=[]
num_string="123456789"
for elemento in permutations(num_string, num_comb):
    permutaciones.append(elemento)

In [3]:
len(permutaciones)

15120

Hay 15120 posibles combinaciones en cuanto a los números se refiere

In [4]:
# Vemos el aspecto del segundo elemento de la lista
permutaciones[2]

('1', '2', '3', '4', '7')

In [5]:
# Ahora genero una lista que contiene en cada elemento una lista con los 
# 4 simbolos operadores correspondientes a las 24 permutaciones posibles a tener
#  en cuenta en la solución.
oper=[]
oper_string="+-*/"
for elemento in permutations(oper_string, len(oper_string)):
    oper.append(elemento)

In [6]:
len(oper)

24

Efectivamente queda comprobado que hay 24 combinaciones diferentes

In [7]:
# Vemos el aspecto del segundo elemento de la lista
oper[11]

('-', '/', '*', '+')

In [8]:
# Ahora creamos strings con cada elemento de permutaciones y con cada uno de 
# estos los 24 posibilidades de la posicion de los operadores, es decir los 
# elementos de la lista oper
# Creamos una lista de soluciones final que contendrá esos 9 caracteres 
# extraídos de dicha combinación:
solucion_final=[]
for perlist in permutaciones:
  for oplist in oper:
    una_sol=''
    for per,op in zip(perlist,oplist):
      una_sol=una_sol+str(per)+op
    # Añadimos el último elemento
    una_sol=una_sol + str(perlist[4])
    solucion_final.append(una_sol)


In [9]:
len(solucion_final)

362880

Ahora que ya tenemos todas las soluciones, utilizamos eval para encontrar el máximo y el mínimo

In [10]:
vmax,vmin,index=0,0,0
# imax=0
# imin=0
for exp in solucion_final:
  if eval(exp)>vmax:
    imax=index
    vmax=eval(exp)
  elif eval(exp)<vmin:
    imin=index
    vmin=eval(exp)
  index+=1

• Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las
siguientes cuestiones:
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?
- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?<br>
• Nota: Es posible usar la función de python “eval” para evaluar una expresión:
....

(*) La respuesta es obligatoria

In [11]:
# Escribimos el valor de la expresión de solución final con el valor máximo
print(solucion_final[imax])

7-1/6+8*9


In [12]:
# Cuyo valor entero máximo es:
print(math.trunc(vmax))

78


In [13]:
# Escribimos el valor de la expresión de solución final con el valor mínimo
print(solucion_final[imin])

1+2/7-8*9


In [14]:
# Cuyo valor entero mínimo es:
print(math.trunc(vmin))

-70


In [15]:
# Escribimos la lista de resultados donde se evalua en cada elemento el valor de la expresión
results = [math.trunc(eval(x)) for x in solucion_final]
# Guardamos los valores únicos eliminando los repetidos en la lista set_resultados
set_results=set(results)

In [16]:
# Vemos si quedan valores entre el máximo y el mínimo que no han podido resolverse mediante alguna expresión de la solucion final
num_elem = -(-70) + 1 + 78 # los negativos, el cero y los positivos
num_elem_sol_final= len(set_results)
print(num_elem, num_elem_sol_final)

149 149


Esto significa, por lo tanto, que en este caso, podemos abarcar el rango de enteros entre -70 y 78 mediante una expresión descrita en el enunciado.

In [17]:
# Ahora de todas las expresiones vamos a seleccionar las 149 que abarcan el 
# resultado [-70,78] y las vamos a guardar el una lista de tuplas que llamaremos 
# tup_list:
tup_list=[]
v_list=[]
for valor, expresion in zip(results, solucion_final):
  if valor not in v_list:
    v_list.append(valor)
    tup_list.append((valor,expresion))

In [18]:
# Volvemos a comprobar que esán los 149 valores con las 149 expresiones
len(tup_list)

149

In [19]:
# Comprobación, ejemplo: supongamos que nos piden que demos una expresion que nos de 59, por ejemplo:
num_a_buscar=59
i=0
found= False
while i<len(tup_list)-1 and not found:
  if tup_list[i][0]==num_a_buscar:
    expresion_a_devolver=tup_list[i][1]
    found=True
  else:
    i+=1
print("La expresión solución es ",expresion_a_devolver, "= ", num_a_buscar)

La expresión solución es  1/2-4+7*9 =  59


(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>



¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.




Si no hubiera habido restricciones, entiendo que se podrian repetir los números y los símbolos operadores por lo que entonces tendríamos 9^9*4^4 expresiones posibles para abarcar el rango de soluciones a evaluar.

Con las restricciones descritas hay una variacion sin repetición seleccionando 5 elementos de 9 para cada combinación de simbolos operadores. En este caso son 15120 variaciones por 24 combinaciones de operaciones que hacen un total de 362880 expresiones.


Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


Mi estructura de datos que se adapta mejor al problema ha sido:
- Una lista donde se han ido guardando strings con las expresiones extraidas de las diferentes combinaciones posibles (solucion_final)
- Una cadena de carácteres o string para cada solución donde se iban añadiendo los caracteres combinando así: <br>
'[entero1][símbolo1][entero2][símbolo2][entero3][símbolo3][entero4][símbolo4][entero5]'<br>
Esta cadena se guardaba en una_sol y se añadía a la lista solucion_final para cada combinación. 

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

En este caso mi objetivo es encontrar todas las expresiones y ver si entre ellas y teniendo en cuenta de que hay un valor máximo y uno mínimo que puedo representar, existen expresiones para cada entero delimitado en dicho intervalo.
En este caso, sí se cumple que hay, al menos una expresión para lograr dicho cometido.<br>
Este problema, pues, no es de maximización ni de optimización ya que es un problema de búsqueda de las soluciones.<br>
Sí habría sido de optimización, por ejemplo si nos hubieran pedido dentro de un intervalo (max,mín) encontrar las expresiones que definan dicho espacio. En dicho caso, no se trataría de optimizar como tal sino de encontrar la manera más elegante y de menos coste para encontrar dichas expresiones sin tener que recorrer todo el espacio de soluciones. En este caso, primero había que ver TODAS las soluciones.

Diseña un algoritmo para resolver el problema por fuerza bruta

Lo metemos todo en una función a la que se le meta un número (n), un número de cantidad de números a combinar (num_comb), una cadena de caracteres con los elementos de donde combinar (num_string) y por último un string con los operadores posibles (oper_string) y, si el número está dentro del intervalo de soluciones posibles, te devuelve una expresión que lo resuelve.

In [20]:
def alg_expresion_numero(n,num_comb,num_string, oper_string):
  # Genero una lista que contiene en cada elemento una lista con 5 números 
  # correspondientes a las permutaciones posibles a tener en cuenta en la 
  # solución.
  permutaciones=[]

  for elemento in permutations(num_string, num_comb):
      permutaciones.append(elemento)

  # Ahora genero una lista que contiene en cada elemento una lista con los 
  # 4 simbolos operadores correspondientes a las 24 permutaciones posibles a tener
  #  en cuenta en la solución.
  oper=[]

  for elemento in permutations(oper_string, len(oper_string)):
      oper.append(elemento)

  # Ahora creamos strings con cada elemento de permutaciones y con cada uno de 
  # estos los 24 posibilidades de la posicion de los operadores, es decir los 
  # elementos de la lista oper
  # Creamos una lista de soluciones final que contendrá esos 9 caracteres 
  # extraídos de dicha combinación:
  solucion_final=[]
  for perlist in permutaciones:
    for oplist in oper:
      una_sol=''
      for per,op in zip(perlist,oplist):
        una_sol=una_sol+str(per)+op
      # Añadimos el último elemento
      una_sol=una_sol + str(perlist[-1])
      solucion_final.append(una_sol)

  # Escribimos la lista de resultados donde se evalua en cada elemento el valor de la expresión
  resultados = [int(math.trunc(eval(x))) for x in solucion_final]

  # Guardamos los valores únicos eliminando los repetidos en la lista set_resultados
  set_resultados=set(resultados)

  # Ahora de todas las expresiones vamos a seleccionar las que abarcan el espacio
  # de soluciones y las vamos a guardar el una lista de tuplas que llamaremos 
  # tup_list:
  tup_list=[]
  v_list=[]
  for valor, expresion in zip(resultados, solucion_final):
    if valor not in v_list:
      v_list.append(valor)
      tup_list.append((valor,expresion))

  # Por último construimos el buscador para un número n de entrada la respuesta 
  # solución:
  min_resul= min(resultados)
  max_resul= max(resultados)
  if n in list(range(min_resul, max_resul+1)):
    found= False
    i=0
    while i<len(tup_list)-1 and not found:
      if tup_list[i][0]==n:
        expresion_a_devolver=tup_list[i][1]
        found=True
      else:
        i+=1
    expresion_return="La expresión solución es " + expresion_a_devolver + "= " + str(n)
  else:
    expresion_return="El número a lograr mediante las condiciones especificadas en el \
    enunciado debe de estar entre " + str(min_resul) + "y " + str(max_resul) + "!!!"
  
  return expresion_return


Calcula la complejidad del algoritmo por fuerza bruta

Debido al triple for anidado al principio del algoritmo al crear la lista de soluciones final, la complejidad de este algoritmo sería n^3 pero debido a la función permutations de itertools la complejidad asciende a O(n.n!).

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

No he encontrado uno que mejore la complejidad ya que sí o sí debo realizar todas las permutaciones para encontrar las soluciones y eso no cambia, siempre es O(n.n!).

(*)Calcula la complejidad del algoritmo 

Por las condiciones del problema siempre es O(n.n!)

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

In [28]:
# Generamos 5 numeros aleatorios entre el -70 y el 78 y los colocamos en una lista. 
rand_list=[]
lower_bound = -70
upper_bound = 78
number_of_random_numbers = 5

for i in range(number_of_random_numbers):
    random_number = random.randint(lower_bound, upper_bound)
    rand_list.append(random_number)


Aplica el algoritmo al juego de datos generado

In [29]:
# Aplicamos el algoritmo a los 5 valores encontrados aleatoriamente
for rand_num in rand_list:
  print(alg_expresion_numero(rand_num,5,'123456789', '+-*/'))


La expresión solución es 1/2-3*5+7= -7
La expresión solución es 1/2-4*9+8= -27
La expresión solución es 1-2/3+5*6= 30
La expresión solución es 1/2-4+5*6= 26
La expresión solución es 1+2/3-5*7= -33


Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

https://www.delftstack.com/es/howto/python/python-max-value-in-list/
https://blog.finxter.com/how-to-convert-an-integer-list-to-a-string-list-in-python/
https://www.w3schools.com/jsref/jsref_trunc.asp
https://docs.python.org/es/3/library/itertools.html
http://recursostic.educacion.es/descartes/web/materiales_didacticos/Combinatoria/variacionessin.htm#:~:text=%C2%BFQu%C3%A9%20son%3F,representa%20por%20Vm%2Cn.


Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Aquí lo interesante sería poder crear un algoritmo que aceptase repeticiones tanto en los simbolos como en los números. Esto generaría un espacio de "resultados" mucho más amplio debido a una lista de soluciones final mucho mas numerosa y por tanto el tiempo de ejecución aumentaría. 

En este caso, tampoco sabemos si todos los números comprendidos en el intervalo máximi y mínimo estarían representados, por lo que también sería interesante para que dimensiones y combinaciones el problema tendría "cobertura total" de los números en el interior del intervalo.